core/src/program/blocks/span_block.rs

```diff
@@ -16,25 +16,29 @@ const MAX_OPS_PER_BATCH: usize = GROUP_SIZE * BATCH_SIZE;
 
 // SPAN BLOCK
 // ================================================================================================
-/// A code block used to describe a linear sequence of operations (i.e., no branching or loops).
+/// Block composed of a sequence of operations.
 ///
-/// When the VM executes a Span block, it breaks the sequence of operations into batches and
-/// groups according to the following rules:
-/// - A group may contain up to 9 operations or a single immediate value.
-/// - An operation carrying an immediate value cannot be in the 9th position in a group.
-/// - A batch may contain up to 8 groups.
-/// - There is no limit on the number of batches contained within a single span.
+/// Executes its operations in order. Fails if any of the operations fails.
 ///
-/// Thus, for example, executing 8 pushes in a row will result in two operation batches:
-/// - The first batch will contain 8 groups, with the first group containing 7 push opcodes,
-///   and the remaining 7 groups containing immediate values for each of the push operations.
-/// - The second batch will contain 2 groups, with the first group containing a single push opcode,
-///   and the second group containing the immediate value for the last push operation.
+/// The hash of a span block is:
 ///
-/// If a sequence of operations does not have any operations which carry immediate values, then
-/// up to 72 operations can fit into a single batch.
+/// > hash(batches, domain=SPAN_DOMAIN)
+///
+/// A span may contain an arbitrary number of batches. A batch is composed of exactly 8 groups,
+/// each group is represented by a field element with 64bits, so a batch is 512bits. Each opcode is
+/// represented with 7bits, so up to 9 opcodes can be packed in a group. An immediate value uses
+/// 64bits, so every immediate value is a group. Opcodes with an immediate value can not be packed
+/// last in the group, and the immediate value must be in the same batch as its opcode,
+/// zero-padding is added to satisfy these requirements. Up to 72 operations without immediate
+/// values can fit into a single batch.
+///
+/// Example: 8 pushes result in two operation batches:
+///
+/// - First batch: First group with 7 push opcodes and 2 zero-paddings packed together, followed by
+/// 7 groups with their respective immediate values.
+/// - Second batch: First group with the last push opcode and 8 zero-paddings packed together,
+/// followed by one immediate and 6 padding groups.
 ///
```

Comment:
I think the previous description was more clear - i.e.,:

- I'd still mention that a span block does not contain any control flow instructions.
- I'd keep the bullet-point list for describing the rules. We can re-order it to have the same flow as in your paragraph.

In the example, the part about "zero-paddings" wasn't clear to me. There are actually a few ways to describe it. We could say that there are just 7 opcodes because that's what will get executed by the VM. Or if we want to be exhaustive, we can say that it is 7 opcodes followed by 2 NOOPs (since numeric value of a NOOP is 0), but then we need to clarify that the VM doesn't always have to execute the NOOPs. 

Also minor nits:
- For bullet point lists, if the content splits onto multiple lines I usually add 2 spaces in front of the second and subsequent lines to improve readability a bit.
- I'd get rid of the trailing `///`.